lxd/device/unix_hotplug.go

```diff
@@ -149,29 +149,31 @@ func (d *unixHotplug) Start() (*deviceConfig.RunConfig, error) {
 	runConf := deviceConfig.RunConfig{}
 	runConf.PostHooks = []func() error{d.Register}
 
-	device := d.loadUnixDevice()
-	if d.isRequired() && device == nil {
+	devices := d.loadUnixDevices()
+	if d.isRequired() && devices == nil {
 		return nil, fmt.Errorf("Required Unix Hotplug device not found")
 	}
 
-	if device == nil {
+	if devices == nil {
 		return &runConf, nil
 	}
 
-	devnum := device.Devnum()
-	major := uint32(devnum.Major())
-	minor := uint32(devnum.Minor())
-
-	// setup device
-	var err error
-	if device.Subsystem() == "block" {
-		err = unixDeviceSetupBlockNum(d.state, d.inst.DevicesPath(), "unix", d.name, d.config, major, minor, device.Devnode(), false, &runConf)
-	} else {
-		err = unixDeviceSetupCharNum(d.state, d.inst.DevicesPath(), "unix", d.name, d.config, major, minor, device.Devnode(), false, &runConf)
-	}
+	for _, device := range devices {
+		devnum := device.Devnum()
+		major := uint32(devnum.Major())
+		minor := uint32(devnum.Minor())
+
+		// setup device
+		var err error
+		if device.Subsystem() == "block" {
+			err = unixDeviceSetupBlockNum(d.state, d.inst.DevicesPath(), "unix", d.name, d.config, major, minor, device.Devnode(), false, &runConf)
+		} else {
+			err = unixDeviceSetupCharNum(d.state, d.inst.DevicesPath(), "unix", d.name, d.config, major, minor, device.Devnode(), false, &runConf)
+		}
 
-	if err != nil {
-		return nil, err
+		if err != nil {
+			return nil, err
```

Comment:
Do we need to consider undoing any previously setup devices if one of the subsequent device setups fails here?

Reply:
I.e. doing the equivalent of when the instance is stopped.

Reply:
But ur doing this in a for loop which means we need to undo files created in earlier iterations if we encounter an error and return

Reply:
Gotcha, makes sense!